bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java

```diff
@@ -2455,6 +2518,59 @@ public void setAutoExpandLevel(int level) {
 		expandToLevel = level;
 	}
 
+	/**
+	 * Sets the level of auto expand on paths of consecutive single children.
+	 * <code> NO_EXPAND </code> means that such paths are not automatically
+	 * expanded.
+	 * <p>
+	 * Auto Expansion of paths of consecutive single children is off by default.
+	 * Turning this behaviour off should be done cautiously on trees with
+	 * lazy-loaded child-nodes.
+	 * <p>
+	 * <p>
+	 * Using <code> ALL_LEVELS </code> as arguments will set to infinite
+	 * auto-expansion of paths of consecutive single children.
+	 * </p>
+	 *
+	 * @param level <code> NO_EXTEND </code> for disabled, <code> ALL_LEVELS </code>
+	 *              for infinite expansion or any integer value to set a level to
+	 *              which paths of consecutive single children are automatically
+	 *              expanded to
+	 * @since 3.32
+	 */
+	public void setAutoExpandOnSingleChildLevel(int level) {
+		autoExpandOnSingleChildLevel = level;
+		if (level == NO_EXPAND) {
+			removeAutoExpandOnSingleChildListener();
+		} else {
+			registerAutoExpandOnSingleChildListener();
+		}
+	}
+
+	private void registerAutoExpandOnSingleChildListener() {
+		autoExpandOnSingleChildListener = new ITreeViewerListener() {
+			@Override
+			public void treeCollapsed(TreeExpansionEvent event) {
+				// Do nothing
+			}
+
+			@Override
+			public void treeExpanded(TreeExpansionEvent e) {
+				Widget item = doFindItem(e.getElement());
+
+				internalConditionalExpandToLevel(item, autoExpandOnSingleChildLevel,
+						w -> Boolean.valueOf(doesWidgetHaveExactlyOneChild(w)));
+			}
+		};
+		addTreeListener(autoExpandOnSingleChildListener);
+	}
+
+	private void removeAutoExpandOnSingleChildListener() {
+		if (autoExpandOnSingleChildListener != null) {
```

Comment:
The [method that removes the listener from the list](https://github.com/eclipse-equinox/equinox/blob/66972dd6c7b1e63106505b3cd67d0a034b4c817f/bundles/org.eclipse.equinox.common/src/org/eclipse/core/runtime/ListenerList.java#L221) will throw an `IllegalArgumentException` if the listener is null.
If we want to remove the `null`-check, we have a few options:
- Allow the method throw in case the listener is null and make the user handle this: I really don't like this option since this exposes internal state to the user. A trivial way to reach this state is eg. if the user calls `setAutoExpandOnSingleChildListener(NO_EXPAND)` before anything else - as done in our own unit test [here](https://github.com/eclipse-platform/eclipse.platform.ui/pull/1072/files#diff-82159efdd73bf97fd8fdde7cb0215f0fd2e302f1c87c2c0423d28e644c230837R190)
- Catch the exception ourselves: IMHO, this obfuscates meaning more than it helps.